lib/ocpp/common/websocket/websocket_tls_tpm.cpp

```diff
@@ -0,0 +1,808 @@
+    if (!this->initialized()) {
+        EVLOG_error << "Message received but TLS websocket has not been correctly initialized. Discarding message.";
+        return;
+    }
+
+    std::string message(reinterpret_cast<char*>(msg), len);
+    EVLOG_info << "Received message over TLS websocket: " << message;
+
+    this->message_callback(message);
+}
+
+static bool send_internal(lws* wsi, WebsocketMessage* msg) {
+    static std::vector<char> buff;
+
+    std::string& message = msg->payload;
+    size_t message_len = message.length();
+    size_t buff_req_size = message_len + LWS_PRE;
+
+    if (buff.size() < buff_req_size)
+        buff.resize(buff_req_size);
+
+    // Copy data in send buffer
+    memcpy(&buff[LWS_PRE], message.data(), message_len);
+
+    // TODO (ioan): if we require certain sending over the wire,
+    // we'll have to send chunked manually something like this:
+
+    // Ethernet MTU: ~= 1500bytes
+    // size_t BUFF_SIZE = (MTU * 2);
+    // char *buff = alloca(LWS_PRE + BUFF_SIZE);
+    // memcpy(buff, message.data() + already_written, BUFF_SIZE);
+    // int flags = lws_write_ws_flags(proto, is_start, is_end);
+    // already_written += lws_write(wsi, buff + LWS_PRE, BUFF_SIZE - LWS_PRE, flags);
+
+    auto sent = lws_write(wsi, reinterpret_cast<unsigned char*>(&buff[LWS_PRE]), message_len, msg->protocol);
+
+    if (sent < 0) {
+        // Fatal error, conn closed
+        EVLOG_error << "Error sending message over TLS websocket conn closed.";
+        return false;
+    }
+
+    if (sent < message_len) {
+        EVLOG_error << "Error sending message over TLS websocket. Sent bytes: " << sent
+                    << " Total to send: " << message_len;
+        return false;
+    }
+
+    return true;
+}
+
+void WebsocketTlsTPM::on_writable() {
+    if (!this->initialized() || !this->m_is_connected) {
+        EVLOG_error << "Message sending but TLS websocket has not been correctly initialized/connected.";
+        return;
+    }
+
+    if (message_queue.size() > 0) {
+        EVLOG_info << "Client writable, sending data";
+
+        if (auto* data = conn_data.get()) {
+            if (data->get_state() == EConnectionState::FINALIZED) {
+                EVLOG_error << "Trying to write message to finalized state!";
+                return;
+            }
+
+            std::unique_ptr<WebsocketMessage> message;
+
+            {
+                std::lock_guard<std::mutex> lock(this->queue_mutex);
+                message = std::move(message_queue.front());
+                message_queue.pop();
+            }
+
+            send_internal(data->get_conn(), message.get());
+        }
+    }
+}
+
+void WebsocketTlsTPM::request_write() {
+    if (this->m_is_connected) {
+        if (auto* data = conn_data.get()) {
+            if (data->get_conn())
+                lws_callback_on_writable(data->get_conn());
+        }
+    }
+}
+
+bool WebsocketTlsTPM::send(const std::string& message) {
```

Comment:
Question: should this function be blocking until the message has been sent over the wire? At the moment libwebsockets copies internally the message, and sends it whenever it has the time.